### PR TITLE
add ocl

### PR DESCRIPTION
**Pull request by looooo**
[opencamlib](https://github.com/aewallin/opencamlib)

Checklist

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages
- [x] Build number is 0
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by looooo** [comment]
currently the build is failing because of several errors similar to this:

```
CMake Error at src/cxxlib/cxxlib.cmake:5 (add_subdirectory):
  The binary directory
    C:/bld/opencamlib_1559213820030/work/build/src/cutters
  is already used to build a source directory.  It cannot be used to build
  source directory
    C:/bld/opencamlib_1559213820030/work/src/cutters
  Specify a unique binary directory name.
Call Stack (most recent call first):
  src/CMakeLists.txt:254 (include)
```

Best to fix this directly in ocl. @aewallin. Seems like a sub-directory is added twice?

**Comment by looooo** [comment]
this seems like an linking issue: (windows and osx)

```
FAILED: ocl.so 
: && $BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-clang++ -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -std=c++14 -fmessage-length=0 -I$PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/opencamlib-2018.08 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix  -O3 -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk -mmacosx-version-min=10.9 -bundle -Wl,-headerpad_max_install_names -Wl,-pie -Wl,-headerpad_max_install_names -Wl,-dead_strip_dylibs -Wl,-rpath,$PREFIX/lib -L$PREFIX/lib -o ocl.so CMakeFiles/ocl.dir/pythonlib/ocl_cutters.cpp.o CMakeFiles/ocl.dir/pythonlib/ocl_geometry.cpp.o CMakeFiles/ocl.dir/pythonlib/ocl_algo.cpp.o CMakeFiles/ocl.dir/pythonlib/ocl_dropcutter.cpp.o CMakeFiles/ocl.dir/pythonlib/ocl.cpp.o  common/libocl_common.a dropcutter/libocl_dropcutter.a cutters/libocl_cutters.a geo/libocl_geo.a algo/libocl_algo.a $PREFIX/lib/libpython3.7m.dylib && :
ld: warning: -pie being ignored. It is only used when linking a main executable
Undefined symbols for architecture x86_64:
  "boost::python::instance_holder::deallocate(_object*, void*)", referenced from:
      boost::python::objects::make_holder<0>::apply<boost::python::objects::value_holder<ocl::MillingCutter_py>, boost::mpl::vector0<mpl_::na> >::execute(_object*) in ocl_cutters.cpp.o
      boost::python::objects::make_holder<0>::apply<boost::python::objects::value_holder<ocl::CylCutter>, boost::mpl::vector0<mpl_::na> >::execute(_object*) in ocl_cutters.cpp.o
      boost::python::objects::make_holder<2>::apply<boost::python::objects::value_holder<ocl::CylCutter>, boost::mpl::vector2<double, double> >::execute(_object*, double, double) in ocl_cutters.cpp.o
      boost::python::objects::make_holder<0>::apply<boost::python::objects::value_holder<ocl::BallCutter>, boost::mpl::vector0<mpl_::na> >::execute(_object*) in ocl_cutters.cpp.o
      boost::python::objects::make_holder<2>::apply<boost::python::objects::value_holder<ocl::BallCutter>, boost::mpl::vector2<double, double> >::execute(_object*, double, double) in ocl_cutters.cpp.o
      boost::python::objects::make_holder<0>::apply<boost::python::objects::value_holder<ocl::BullCutter>, boost::mpl::vector0<mpl_::na> >::execute(_object*) in ocl_cutters.cpp.o
      boost::python::objects::make_holder<3>::apply<boost::python::objects::value_holder<ocl::BullCutter>, boost::mpl::vector3<double, double, double> >::execute(_object*, double, double, double) in ocl_cutters.cpp.o
```

linux is failing earlier with a pedantic error:

```
FAILED: CMakeFiles/ocl.dir/pythonlib/ocl_geometry.cpp.o 
$BUILD_PREFIX/bin/x86_64-conda_cos6-linux-gnu-c++  -Docl_EXPORTS -I. -I$PREFIX/include/python3.6m -I$SRC_DIR/src/cutters -I$SRC_DIR/src/geo -I$SRC_DIR/src/algo -I$SRC_DIR/src/dropcutter -I$SRC_DIR/src/common -I$SRC_DIR/src -fvisibility-inlines-hidden -std=c++17 -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -I$PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/opencamlib-2018.08 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -fopenmp -O3 -DNDEBUG -fPIC   -Wall -Wno-deprecated -pedantic-errors -std=gnu++11 -MD -MT CMakeFiles/ocl.dir/pythonlib/ocl_geometry.cpp.o -MF CMakeFiles/ocl.dir/pythonlib/ocl_geometry.cpp.o.d -o CMakeFiles/ocl.dir/pythonlib/ocl_geometry.cpp.o -c $SRC_DIR/src/pythonlib/ocl_geometry.cpp
In file included from $PREFIX/include/boost/python/object/function_handle.hpp:8:0,
                 from $PREFIX/include/boost/python/converter/arg_to_python.hpp:19,
                 from $PREFIX/include/boost/python/call.hpp:15,
                 from $PREFIX/include/boost/python/object_core.hpp:14,
                 from $PREFIX/include/boost/python/args.hpp:22,
                 from $PREFIX/include/boost/python.hpp:11,
                 from $SRC_DIR/src/pythonlib/ocl_geometry.cpp:21:
$PREFIX/include/boost/python/detail/caller.hpp:127:2: error: extra ';' [-Wpedantic]
```

File: recipes/ocl/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+package:
+    name: {{ name }}
+    version: {{ version }}
+
+source:
+    # fn: {{ name }}.tar.gz
+    # url: https://github.com/aewallin/{{ name }}/archive/{{ version }}.tar.gz
+    # sha256: {{ sha256 }}
+    git_url: https://github.com/aewallin/opencamlib
+    git_tag: master
```
Comment (marcelotrevisani) [inline recipes/ocl/meta.yaml]: Is it possible to use the tarballs instead of the ``git_url``?

File: recipes/ocl/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        - boost 1.68
+        - python
+
+test:
+  imports:
+    - ocl
+
+about:
+    home: http://www.anderswallin.net/CAM/
+    license: GPL3.0
```
Comment (marcelotrevisani) [inline recipes/ocl/meta.yaml]: Could you please change it to: ``GPL-3.0`` ?

File: recipes/ocl/meta.yaml (hunk 1 above)
Comment (looooo) [reply recipes/ocl/meta.yaml]: I am waiting for a release which hopefully comes.

File: recipes/ocl/meta.yaml (hunk 2 above)
Comment (looooo) [reply recipes/ocl/meta.yaml]: the license changed to  LGPL-2.1.
https://github.com/aewallin/opencamlib/blob/master/COPYING

**Comment by looooo** [comment]
@conda-forge/core was there a recent update of conda-build? Maybe you can have a short look at these errors, not sure if this is related to modified sources.

osx:

```
undefined symbols for architecture x86_64:
  "boost::python::instance_holder::deallocate(_object*, void*)",
```

linux:
```
CMake Error at cmake_install.cmake:77 (file):
  file INSTALL cannot set permissions on
  "/opt/conda/lib/python3.7/site-packages"
```

win:

works, whaaaaat?!?

**Comment by looooo** [comment]
seems like on linux a wrong python is found (from base-environment).

**Comment by looooo** [comment]
@conda-forge/core I am not sure if this is a problem with conda-build. But some time ago this recipe worked and now python-detection seems to be broken for linux and py36. py37 seems to work.
The python-detection is done here:
https://github.com/aewallin/opencamlib/blob/master/src/pythonlib/pythonlib.cmake#L57
and result in (py36):
```
-- PYTHON_SITE_PACKAGES = /opt/conda/lib/python3.7/site-packages
-- PYTHON_ARCH_PACKAGES = /opt/conda/lib/python3.7/site-packages
-- PYTHON_INCLUDE_DIRS = /opt/conda/include/python3.7m
-- PYTHON_LIBRARIES = /opt/conda/lib/libpython3.7m.so
-- PYTHON_SITE_PACKAGES = /opt/conda/lib/python3.7/site-packages
-- PYTHON_ARCH_PACKAGES = /opt/conda/lib/python3.7/site-packages
```

which is obviously wrong. As the problem only occures with python36 I guess this problem is not related to this library but some changed conda-libraries (conda-build). Not sure how to further investigate. For now I disabled py36.

**Comment by scopatz** [comment]
Do you have a link to the failing build?

**Comment by looooo** [comment]
@scopatz py36 builds are still failing: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=58724&view=logs

File: recipes/ocl/build.sh [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+if [[ ${HOST} =~ .*linux.* ]]; then
+	# https://github.com/conda-forge/boost-feedstock/issues/72
+	sed -i '127s#;##g' ${PREFIX}/include/boost/python/detail/caller.hpp
+fi
+
+cmake -G "Ninja" \
+      -D CMAKE_BUILD_TYPE:STRING=Release \
+      -D CMAKE_PREFIX_PATH:FILEPATH=$PREFIX \
+      -D CMAKE_INSTALL_PREFIX:FILEPATH=$PREFIX \
```
Comment (scopatz) [inline recipes/ocl/build.sh]: Try putting the above in quotes, either `CMAKE_INSTALL_PREFIX:FILEPATH="${PREFIX}"` or `"CMAKE_INSTALL_PREFIX:FILEPATH=${PREFIX}"`. Also try removing `":FILEPATH"` from these.

Comment (looooo) [reply recipes/ocl/build.sh]: Hmm, as far as I can remember this was never a problem. But I will try.

Comment (looooo) [reply recipes/ocl/build.sh]: ok,  problem is related to cmake 3.15 (worked previously with cmake 3.14.)

cmake 3.15 breaks
```
find_package(Python3 ...)
```
for conda environments

**Comment by looooo** [comment]
:tada: ready to be merged.

File: recipes/ocl/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+build:
+    number: 0
+    skip: true  # [py<36]
+
+requirements:
+    build:
+        - {{ compiler("cxx") }}
+        - ninja
+        - cmake 3.14
```
Comment (scopatz) [inline recipes/ocl/meta.yaml]: can this be cmake >=3.14

Comment (looooo) [reply recipes/ocl/meta.yaml]: No, see referenced issue: https://github.com/conda-forge/cmake-feedstock/issues/91

Comment (scopatz) [reply recipes/ocl/meta.yaml]: Well what about <=3.14? or !=3.15?

Comment (looooo) [reply recipes/ocl/meta.yaml]: Ok, this should work.

File: recipes/ocl/build.sh [hunk 5]
```diff
@@ -0,0 +1,19 @@
+mkdir build -p
+cd build 
+
+if [[ ${HOST} =~ .*linux.* ]]; then
```
Comment (isuruf) [inline recipes/ocl/build.sh]: Use `target_platform` variable here

Comment (looooo) [reply recipes/ocl/build.sh]: thanks, done.

Comment (looooo) [reply recipes/ocl/build.sh]: btw.: I didn't find this in the docs. Should it be added?

Comment (isuruf) [reply recipes/ocl/build.sh]: HOST is going to be renamed soon. I can't find the discussion at the moment

Comment (scopatz) [reply recipes/ocl/build.sh]: hmm weird.  Yeah, I missed that too. Docs would be great for that

**Comment by scopatz** [comment]
Thanks @looooo!

**Comment by isuruf** [comment]
Found the discussion https://github.com/conda/conda/issues/7031

File: recipes/ocl/build.sh (hunk 5 above)
Comment (jakirkham) [reply recipes/ocl/build.sh]: Copying @isuruf's comment here for continuity.

> Found the discussion [conda/conda#7031](https://github.com/conda/conda/issues/7031)

**Pull request merged**